flink-connector-obkv-hbase/src/main/java/com/oceanbase/connector/flink/sink/OBKVHBaseStatementExecutor.java

```diff
@@ -0,0 +1,150 @@
+
+import com.oceanbase.connector.flink.connection.OBKVHBaseConnectionProvider;
+import com.oceanbase.connector.flink.connection.OBKVHBaseTableSchema;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.util.Bytes;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+public class OBKVHBaseStatementExecutor implements StatementExecutor<RowData> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final OBKVHBaseStatementOptions options;
+    private final OBKVHBaseTableSchema tableSchema;
+    private final OBKVHBaseConnectionProvider connectionProvider;
+
+    private final List<RowData> buffer = new ArrayList<>();
+
+    private SinkWriterMetricGroup metricGroup;
+
+    public OBKVHBaseStatementExecutor(
+            OBKVHBaseStatementOptions options,
+            OBKVHBaseTableSchema tableSchema,
+            OBKVHBaseConnectionProvider connectionProvider) {
+        this.options = options;
+        this.tableSchema = tableSchema;
+        this.connectionProvider = connectionProvider;
+    }
+
+    @Override
+    public void setContext(Sink.InitContext context) {
+        this.metricGroup = context.metricGroup();
+    }
+
+    @Override
+    public void addToBatch(RowData record) {
+        synchronized (buffer) {
+            buffer.add(record);
+        }
+        metricGroup.getIOMetricGroup().getNumRecordsInCounter().inc();
+    }
+
+    @Override
+    public void executeBatch() throws Exception {
+        synchronized (buffer) {
+            int count = 0;
+            List<Put> putList = new LinkedList<>();
+            List<Delete> deleteList = new LinkedList<>();
+            for (RowData record : buffer) {
+                if (record.getRowKind() == RowKind.INSERT
+                        || record.getRowKind() == RowKind.UPDATE_AFTER) {
+                    Put put = getPut(record);
+                    if (put != null) {
+                        putList.add(put);
+                    }
+                } else {
+                    Delete delete = getDelete(record);
+                    if (delete != null) {
+                        deleteList.add(delete);
+                    }
+                }
+                count++;
+                if (count >= options.getBatchSize()) {
+                    if (!putList.isEmpty()) {
+                        connectionProvider.getTable().put(putList);
+                    }
+                    if (!deleteList.isEmpty()) {
+                        connectionProvider.getTable().delete(deleteList);
+                    }
+                    metricGroup.getIOMetricGroup().getNumRecordsOutCounter().inc(count);
+                    count = 0;
+                }
+            }
+            if (count > 0) {
+                connectionProvider.getTable().put(putList);
+                connectionProvider.getTable().delete(deleteList);
```

Comment:
It is more efficient for us to make all operations with the same family and qualifier set into the same putList and deleteList. Put operation using a buffer and sending requests depends on a unique strategy. Delete will directly send requests after you call delete().

> About Put

_HBASE_HTABLE_CLIENT_WRITE_BUFFER_ and _autoFlush_ is important.
- _HBASE_HTABLE_CLIENT_WRITE_BUFFER_: Max buffer size for put operation. Once the heap size reaches the max buffer size, the client will commit(send) all the put operations.
- _autoFlush_: Whether to commit(send) all the put operations after handling a putList(default true).

> About Delete

- Client will do multiple delete operations if you using delete(List<delete>), I guess we could try to make those delete into one delete to reduce the RPC.